module/esami.py

```diff
@@ -61,76 +42,36 @@ def dict_factory(cursor, row):
-		if 'anno' in args:
-			args.remove('anno')
-
-		if 'sessione' in args:
-			args.remove('sessione')
-
-		# Study case
-		if len(args) == 1:
-
-			if args[0] in ("primo", "secondo", "terzo"):
-				output = esami_condition(items, "anno", args[0])
-
-			elif args[0] in ("prima", "seconda", "terza", "straordinaria"):
-				output = esami_condition(items, "sessione", args[0], True)
-
-			elif [item["insegnamento"].lower().find(args[0]) for item in items]:
-				output = esami_condition(items, "insegnamento", args[0])
-
-			output_str = check_output(output)
-
-		elif len(args) > 1:
-
-			# Create an array of session and years if in arguments
-			sessions = list(set(args).intersection(("prima", "seconda", "terza", "straordinaria")))
-			years = list(set(args).intersection(("primo", "secondo", "terzo")))
-
-			_years = {
-				"primo": "1° anno",
-				"secondo": "2° anno",
-				"terzo": "3° anno"
-			}
-			for i in range(len(years)):
-				years[i] = _years[years[i]]
-
-			if sessions and years:
-				for item in items:
-					if (item["anno"] in years) and ([session for session in sessions if [appeal for appeal in item[session] if appeal]]):
-						output.add(esami_output(item, sessions))
-
-			elif sessions and not years:
-				# If years array is empty and session not, the other word are subjects
-				subjects = [arg for arg in args if arg not in(sessions)]
-
-				if subjects:
-					for item in items:
-						for subject in subjects:
-							if [session for session in sessions if [appeal for appeal in item[session] if appeal]] and subject in item["insegnamento"].lower():
-								output.add(esami_output(item, sessions))
-
-			elif not sessions and not years:
-				for arg in args:
-					output = output.union(esami_condition(items, "insegnamento", arg))
-
-			output_str = check_output(output)
-	else:
-		output_str = "Inserisci almeno uno dei seguenti parametri: giorno, materia, sessione (prima, seconda, terza, straordinaria)."
-
-	return output_str
+def esami_cmd(userDict):
+	output_str = []
+ 	#stringa contenente le sessioni per cui la flag è true, separate da ", " 	
+	select_sessione = ", ".join([key for key, value in userDict.items() if userDict.get(key, False) and "sessione" in key]).replace("sessione", "") #=> , prima, seconda, terza
+	#stringa contenente le sessioni per cui la flag è true, separate da " = '[]' and not " 		
+	where_sessione = " = '[]' and not ".join([key for key, value in userDict.items() if userDict.get(key, False) and "sessione" in key]).replace("sessione", "") #=> and not prima = '[] and not terza = '[]' 
+	#stringa contenente gli anni per cui la flag è true, separate da "' and anno = '"	
+	where_anno = "' and anno = '".join([key for key, value in userDict.items() if userDict.get(key, False) and "anno" in key]) #=>  and anno = '1° anno' and anno = '3° anno'
+	#stringa contenente l'insegnamento, se presente		
+	where_insegnamento = userDict.get("insegnamento", "") #=> and insegnamento LIKE '%stringa%'
+
```

Comment:
e' molto bello compattare il codice, pero' a volte si perde di leggibilita' :sweat_smile: .

hai fatto molto bene a mettere i commenti :+1: , pero' comunque il codice sembra un po' disordinato cosi', per me lo puoi lasciare, pero' la prossima volta evita di scrivere tutto in una riga.

Inoltre, se non usi "value", a me sembra ridondante il check che fai con userDict.get(key, False), cioe' da cosi':
```python
	select_sessione = ", ".join([key for key, value in userDict.items() if userDict.get(key, False) and "sessione" in key]).replace("sessione", "") #=> , prima, seconda, terza
```
potresti scrivere direttamente (penso):
```python
	select_sessione = ", ".join([key for key in userDict if "sessione" in key]).replace("sessione", "") #=> , prima, seconda, terza
```

analogamente per le altre righem se vuoi aggiungere il check potresti sempre comunque scrivere `if userDict.get(key, False)` ma comunque eviti di usare .items() e di estrarre il value.
Ma se fai "for key in dict" penso sia inutile fare il check dell'esistenza della chiave, no? :P

Reply:
In realtà il fatto è che la value è utilizzata: dopo aver selezionato un parametro (mettiamo 1° anno) l'utente può rimuoverlo dalla selezione ri-selezionandolo dal menù, e ponendo quindi il valore nella key corrispondente a falso (la key continua ad essere presente). Volendo, si potrebbe semplicemente eliminare la key dal dict, ma avevo optato per questa soluzione per non creare-distruggere troppi riferimenti, ma magari in quel modo diventa più elegante

Reply:
Però in effetti al posto di "userDict.get(key, False)" si potrebbe mettere "value" immagino

Reply:
per un attimo ho pensato che facessi un check sull'esistenza della chiave, invece controlli il valore, si, penso che potresti usare direttamente `value` anziche' usare `dict.get()`.

Alternativamente, puoi evitare di usare dict.items() pero' mantenendo il check che applichi con dict.get(key).

Insomma entrambe queste soluzioni sono equivalenti, ma usare il check in entrambi allunga quella riga compattata piena di codice, quindi se puoi riducila un po' levando il check ed usando value, o levando .items()

```python
asd = { "a": 1, "b": 2 }
for k in asd:
	if asd.get(k):
		print(asd.get(k))

for k, value in asd.items():
	if value is not None:
		print(value)
```

Reply:
Ok, ho semplificato quelle belle righe impegnative sostituendo context.user_data.get(key, False)" => "value", che essendo bool semplifica anche l'if.
Inoltre mi sono anche accorto di un comportamento indesiderato nella composizione della query, e l'ho corretto (la richiesta degli anni era in AND anziché in OR)